components/insight/SRC/org/openmicroscopy/shoola/util/roi/io/ROIReader.java

```diff
@@ -342,58 +342,9 @@ private void setImage(Roi[] rois)
                     }
                 }
             } else if (r instanceof ShapeRoi) {
-                Roi[] subRois = ((ShapeRoi) r).getRois();
-                Roi shapeij;
-                for (int j = 0; j < subRois.length; j++) {
-                    shapeij = subRois[j];
-
-                    // Set ImagePlus reference in subROIs for the check in L216 to work
-                    ImagePlus imp = r.getImage();
-                    shapeij.setImage(imp);
-                    // Transfer correct ROI positions (according to IJ) from superROI
-                    int pos = r.getPosition();
-                    int c = r.getCPosition();
-                    int z = r.getZPosition();
-                    int t = r.getTPosition();
-                    if (imp.getNChannels() == 1 && imp.getNSlices() == 1) {
-                        shapeij.setPosition(pos);
-                    } else if (imp.getNChannels() == 1 &&
-                            imp.getNFrames() == 1) {
-                        shapeij.setPosition(pos);
-                    } else if (imp.getNSlices() == 1 &&
-                            imp.getNFrames() == 1) {
-                        shapeij.setPosition(pos);
-                    } else if (imp.isHyperStack()) {
-                        shapeij.setPosition(c, z, t);
-                    }
-                    type = shapeij.getTypeAsString();
```

Comment:
@bramalingam: I couldn't find a difference between the handling of ROIs and their conversion and the handling of subROIs. Hence, I have refactored the code to a recursive call to `read()` with the subROIs as parameters in L347.

My "testing" workflow properly imports (ImageJ) composite ROIs as individual ROIs to OMERO. Am I missing something completely obvious here, @bramalingam?